Fix canonPath so that driver letters become the root

puffnfresh opened the pull request:
Could be done way better.

I was seeing a lot of `C:\C:\C:\Users\brian` paths being generated, for some reason. Now whenever it sees a path component ending in colon, it just resets.

Ericson2314 committed:
Build a minimized Nix with MinGW

At this point many features are stripped out, but this works:

- Can run libnix{util,store,expr} unit tests
- Can run some Nix commands

Co-Authored-By volth <volth@volth.com>
Co-Authored-By Brian McKenna <brian@brianmckenna.org>

Ericson2314 committed:
Get rid of `shellCrossSystems`

We don't need it now that our (minimized) Windows build of Nix succeeds!

Ericson2314 committed:
Prefix `-DNIX_` paths to be windows-complient for windows

This is a hacky solution, but it will do for now.

Ericson2314 committed:
Relax store path canonicalization

puffnfresh committed:
Fix canonPath so that driver letters become the root

Ericson2314 commented:
Can you make a unit test for this? Unfortunately this sort of fix will not work, because `file-path-impl.hh` is used for `CanonPath` too, which is always Unix-style on all platforms.

I think we can fix this by using [`std::filesystem::weakly_canonical`](https://en.cppreference.com/w/cpp/filesystem/canonical) in the resolve symlinks case, and sticking with something like my version of splitting off the `C:\` first in the non-resolve-symlinks case.

Ericson2314 committed:
Add unix (and linux) dirs

In the Nix commit, platform-specific sources will go here.

Ericson2314 committed:
Start factoring out Unix assumptions

This splits files and adds new identifiers in preperation for supporting
windows, but no Windows-specific code is actually added yet.
